pkg/services/ngalert/state/manager.go

```diff
@@ -113,69 +113,97 @@ func (st *Manager) Warm(ctx context.Context, rulesReader RuleReader) {
 
 	statesCount := 0
 	states := make(map[int64]map[string]*ruleStates, len(orgIds))
-	for _, orgId := range orgIds {
-		// Get Rules
-		ruleCmd := ngModels.ListAlertRulesQuery{
-			OrgID: orgId,
-		}
-		alertRules, err := rulesReader.ListAlertRules(ctx, &ruleCmd)
-		if err != nil {
-			st.log.Error("Unable to fetch previous state", "error", err)
-		}
+	for _, orgID := range orgIds {
+		orgStates, count := st.warmOrg(ctx, orgID, rules)
+		statesCount += count
+		states[orgID] = orgStates
+	}
+	st.cache.setAllStates(states)
+	st.log.Info("State cache has been initialized", "states", statesCount, "duration", time.Since(startTime))
+}
 
-		ruleByUID := make(map[string]*ngModels.AlertRule, len(alertRules))
-		for _, rule := range alertRules {
-			ruleByUID[rule.UID] = rule
-		}
+func (st *Manager) warmOrg(ctx context.Context, orgID int64, rules RuleReader) (map[string]*ruleStates, int) {
```

Comment:
The diff here is nasty, all that's happened is moving the contents of the outer loop over each org into a separate function.

Reviewing commit by commit makes this more clear.